contracts/src/zaps/UniV3Zap.sol

```diff
@@ -0,0 +1,937 @@
+
+        return false;
+    }
+
+    /// @dev Validate the swap parameters for zapping tokens out of Hyperdrive.
+    /// @param _hyperdrive The Hyperdrive pool to open the long on.
+    /// @param _hyperdriveOptions The options that configure how the Hyperdrive
+    ///        operation is settled.
+    /// @param _swapParams The Uniswap swap parameters for a multi-hop fill.
+    /// @param _shouldWrap A flag indicating that the Hyperdrive proceeds should
+    ///        be wrapped before the swap is made.
+    /// @return The asset that will be output by the Hyperdrive operation as
+    ///         proceeds.
+    function _validateZapOut(
+        IHyperdrive _hyperdrive,
+        IHyperdrive.Options calldata _hyperdriveOptions,
+        ISwapRouter.ExactInputParams memory _swapParams,
+        bool _shouldWrap
+    ) internal view returns (address) {
+        // Ensure that the swap recipient is the sender.
+        if (_hyperdriveOptions.destination != address(this)) {
+            revert InvalidRecipient();
+        }
+
+        // If the withdrawal is in base, the proceeds are in the base token.
+        // Otherwise, they are in the vault shares token.
+        address proceedsAsset;
+        if (_hyperdriveOptions.asBase) {
+            proceedsAsset = _hyperdrive.baseToken();
+        } else {
+            proceedsAsset = _hyperdrive.vaultSharesToken();
+        }
+
+        // Validate the zap if the proceeds are in ETH.
+        address tokenIn = _swapParams.path.tokenIn();
+        if (proceedsAsset == ETH) {
+            // Ensure that we are wrapping the proceeds.
+            if (!_shouldWrap) {
+                revert ShouldWrapAssets();
+            }
+
+            // Ensure that the input token is WETH.
+            if (tokenIn != address(weth)) {
+                revert InvalidInputToken();
+            }
+        }
+        // Validate the other cases.
+        else {
+            // Ensure that if we aren't wrapping the proceeds that the proceeds
+            // asset matches the input token.
+            if (!_shouldWrap && proceedsAsset != tokenIn) {
+                revert InvalidInputToken();
+            }
+
+            // Ensure that if we are wrapping the proceeds that the proceeds
+            // asset doesn't match the input token.
+            if (_shouldWrap && proceedsAsset == tokenIn) {
+                revert InvalidInputToken();
+            }
+        }
+
+        return proceedsAsset;
+    }
+
+    /// @dev Zaps funds into this contract to open positions on Hyperdrive.
+    /// @param _zapInOptions The options for the zap.
+    /// @param _shouldConvertToETH A flag indicating whether or not the proceeds
+    ///        should be converted to ETH.
+    /// @return proceeds The amount of assets that were zapped into this
+    ///         contract.
+    function _zapIn(
+        IUniV3Zap.ZapInOptions memory _zapInOptions,
+        bool _shouldConvertToETH
+    ) internal returns (uint256 proceeds) {
```

Comment:
what do we do if tokenIn==tokenOut?

Reply:
Good question. My approach so far was "do nothing and let Uniswap handle it," but I can add an explicit error and failure for this for zapping in and out.